### PR TITLE
Add @orama/plugin-pt15

### PR DESCRIPTION
**Pull request by allevo**
Implement different index algorithm.

### CODE REVIEW

File: packages/plugin-pt15/src/algorithm.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
+import { AnyIndexStore, AnyOrama, SearchableType, Tokenizer, VectorIndex } from "@orama/orama"
+import { avl, bkd, flat, bool } from '@orama/orama/trees'
+import {
+  getVectorSize, internalDocumentIDStore, isVectorType } from '@orama/orama/components'
+
+type InternalDocumentID = internalDocumentIDStore.InternalDocumentID;
+
+export type TreeType = 'AVL' | 'Radix' | 'Bool' | 'Flat' | 'BKD';
```
Comment (micheleriva) [inline packages/plugin-pt15/src/algorithm.ts:8]: We should export these from the `@orama/orama` package

File: packages/plugin-pt15/src/algorithm.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+export type TTree<T = TreeType, N = unknown> = {
```
Comment (micheleriva) [inline packages/plugin-pt15/src/algorithm.ts:9]: This one too

File: packages/plugin-pt15/src/algorithm.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    type: T;
+    node: N;
+    isArray: boolean;
+};
+export type Tree =
```
Comment (micheleriva) [inline packages/plugin-pt15/src/algorithm.ts:14]: This one too

File: packages/plugin-pt15/src/algorithm.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  searchableProperties: string[]
+  searchablePropertiesWithTypes: Record<string, SearchableType>
+}
+
+function create_obj() {
+  // object with empty prototype to cheap objects
+  return Object.create(null)
+}
+
+export function recursiveCreate<T extends AnyOrama>(indexDatastore: PT15IndexStore, schema: T['schema'], prefix: string) {
```
Comment (micheleriva) [inline packages/plugin-pt15/src/algorithm.ts:56]: Is recursion well supported in that case? Should we opt for an iterative approach for slightly larger datasets (50k elements)?

File: packages/plugin-pt15/src/algorithm.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+    if (isVectorType(type)) {
+      indexDatastore.searchableProperties.push(path)
+      indexDatastore.searchablePropertiesWithTypes[path] = type
+      indexDatastore.vectorIndexes[path] = {
+        size: getVectorSize(type),
+        vectors: {}
+      }
+    } else {
+      const isArray = /\[/.test(type as string)
```
Comment (micheleriva) [inline packages/plugin-pt15/src/algorithm.ts:74]: we may want to use `.endsWith("[]")` or even `type[type.length-1] == ']'` for performance reasons

File: packages/plugin-pt15/src/algorithm.ts [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    }
+  }
+}
+
+export function insertString(
+  value: string,
+  positionsStorage: PositionsStorage,
+  prop: string,
+  internalId: InternalDocumentID,
+  language: string | undefined,
```
Comment (micheleriva) [inline packages/plugin-pt15/src/algorithm.ts:126]: I think I started exposing a `Optional` type which is basically `type Optional<T = unknown> = T | undefined`

**Review by micheleriva — APPROVED** [review]
minor comments, but lgtm!

File: packages/plugin-pt15/src/algorithm.ts (hunk 4 above)
Comment (allevo) [reply packages/plugin-pt15/src/algorithm.ts:56]: This is not an intensive path: this function is invoked only during the orama creation

**Pull request merged**